app/controllers/lottery_controller.rb

```diff
@@ -12,64 +12,103 @@ def action_allowed?
-    end
-    bidding_data = {users: priority_info, max_team_size: assignment.max_team_size}
-    ExpertizaLogger.info LoggerMessage.new(controller_name, session[:user].name, "Bidding data for assignment #{assignment.name}: #{bidding_data}", request)
-    url = WEBSERVICE_CONFIG["topic_bidding_webservice_url"]
+
+    user_bidding_info = construct_user_bidding_info(assignment.sign_up_topics, teams)
+    bidding_data = {users: user_bidding_info, max_team_size: assignment.max_team_size}
+    log("Bidding data for assignment #{assignment.name}: #{bidding_data}")
+
     begin
+      url = WEBSERVICE_CONFIG["topic_bidding_webservice_url"]
       response = RestClient.post url, bidding_data.to_json, content_type: :json, accept: :json
+      # Structure of teams variable: [[user_id1, user_id2], [user_id3, user_id4]]
       teams = JSON.parse(response)["teams"]
-      ExpertizaLogger.info LoggerMessage.new(controller_name, session[:user].name, "Team formation info for assignment #{assignment.name}: #{teams}", request)
-      create_new_teams_for_bidding_response(teams, assignment, priority_info)
+      log("Team formation info for assignment #{assignment.name}: #{teams}")
+
+      create_new_teams_for_bidding_response(teams, assignment, user_bidding_info)
+      remove_empty_teams(assignment)
       match_new_teams_to_topics(assignment)
     rescue StandardError => e
       flash[:error] = e.message
     end
+
     redirect_to controller: 'tree_display', action: 'list'
   end
 
   private
 
-  def create_new_teams_for_bidding_response(teams, assignment, priority_info)
-    # Structure of teams variable: [[user_id1, user_id2], [user_id3, user_id4]]
+  # Put generated data to ExpertizaLogger.info
+  def log(message)
+    ExpertizaLogger.info LoggerMessage.new(controller_name, session[:user].name, message, request)
+  end
+
+  # Generate user bidding infomation hash based on students who haven't signed up yet
+  # Structure of user_bidding_info variable: [{user_id1, bids_1}, {user_id2, bids_2}]
+  def construct_user_bidding_info(sign_up_topics, teams)
+    user_bidding_info = []
+    # Exclude any teams already signed up
+    teams_not_signing_up = teams.reject {|team| SignedUpTeam.where(team_id: team.id, is_waitlisted: 0).any? }
+    teams_not_signing_up.each do |team|
+      bids = [0]
+      sign_up_topics.each do |topic|
+        bid_record = Bid.find_by(team_id: team.id, topic_id: topic.id)
+        bids << bid_record.priority if bid_record.try(:priority)
+      end
+      # Grab student id and list of bids
+      team.users.each {|user| user_bidding_info << {pid: user.id, ranks: bids} } if bids.uniq != [0]
+    end
+    user_bidding_info
+  end
+
+  # Generate team bidding infomation hash based on newly-created teams
+  # Structure of team_bidding_info variable: [{team_id1, bids_1}, {team_id2, bids_2}]
+  def construct_team_bidding_info(unassigned_teams, sign_up_topics)
+    team_bidding_info = []
```

Comment:
I would change this variable name to something like bidding_info_for_teams to make it clear that this is bidding info for all unassigned teams rather than just a single team.

Reply:
or maybe teams_bidding_info_for_topics